cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/model/Token.java

```diff
@@ -1,4 +1,144 @@
+    private ZnElement encryptionSecretKey; // secret key used for the ElGamal encryption in the Spend algorithm
+    private ZnElement doubleSpendRandomness0; // randomness used for the first challenge generation in double spending protection
+    private ZnElement doubleSpendRandomness1; // randomness used for the second challenge generation in double spending protection
+    private ZnElement points; // number of points that the token currently stores (initially 0)
+    private ZnElement z, t; // values for blinding the token group element
+
+    /**
+     * standard constructor, intializes an empty token, meaning a token storing 0 points represented as a Pedersen commitment.
+     * Note that certificate is not set since a token is usually not certified upon creation.
+     * @param myRemClassRing remainder class ring of exponents
+     * @param sigScheme signature scheme
+     * @param vKey verification key
+     * @param h array of the bases for the Pedersen commitment
+     * @param USK secret key of user
+     * @param esk ElGamal secret key
+     * @param dsrnd0 double spending randomness
+     * @param dsrnd1 other double spending randomness
+     */
+    public Token(Zn myRemClassRing, SPSEQSignatureScheme sigScheme, VerificationKey vKey,GroupElement[] h, ZnElement USK, ZnElement esk, ZnElement dsrnd0, ZnElement dsrnd1, ZnElement z, ZnElement t) throws PedersenException
+    {
+        // asserting correct number of group elements passed
+        if(h.length != 7)
+        {
+            throw new PedersenException("h is required to consist of 7 group elements, found: " + String.valueOf(h.length));
+        }
+
+        // TODO check out group element vector, vector exponentiation operation
+        // initializing bases array with deep copy of passed elements array
+        this.commitmentBases = new GroupElement[h.length];
+        for(int i = 0; i < h.length; i++)
+        {
+            this.commitmentBases[i] = h[i]; // note: due to 0-based indexing, the bases' indices are off by one wrt paper
+        }
+
+        // initializing respective object variables with other parameters
+        this.myRemainderClassRing = myRemClassRing;
+        this.signatureScheme = sigScheme;
+        this.verificationKey = vKey;
+        this.userSecretKey = USK;
+        this.encryptionSecretKey = esk;
+        this.doubleSpendRandomness0 = dsrnd0;
+        this.doubleSpendRandomness1 = dsrnd1;
+        this.points = myRemainderClassRing.getZeroElement(); // initially, token contains no points
+        this.z = z;
+        this.t = t;
+
+        // compute the token-representing Pedersen commitment for the first time
+        this.updateToken();
+    }
+
+    /**
+     * recomputes the Pedersen commitment representing the token from the group elements and exponents (and also updates its plaintext representation)
+     * (see chapter 4 of 2020 incentive systems paper)
+     */
+    public void updateToken()
```

Comment:
This looks more like a private method to me. Nobody from the outside should call this, right?

Reply:
Yes, thanks. My bad.